File: src/leetcode/python/two_pointers/ThreeSum_15.py
class ThreeSum_15(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        result = []
        nums.sort()
        for i in range(len(nums) - 2):
            if nums[i] > 0:
                break
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            lo, hi = i + 1, len(nums) - 1
            while lo < hi:
                subsum = nums[lo] + nums[hi]
                if subsum > -nums[i]:
                    hi -= 1
                elif subsum < -nums[i]:
                    lo += 1
                else:
                    result.append([nums[i], nums[lo], nums[hi]])
                    lo += 1
                    while lo < hi and nums[lo] == nums[lo - 1]:
                        lo += 1
                    hi -= 1
                    while lo < hi and nums[hi] == nums[hi + 1]:
                        hi -= 1
        return result

File: src/leetcode/java/ThreeSum_15.java
package leetcode.java;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

public class ThreeSum_15 {
  public List<List<Integer>> threeSum(int[] nums) {
    List<List<Integer>> result = new LinkedList<>();
    Arrays.sort(nums);
    for (int i = 0; i < nums.length - 2; i++) {
      if (nums[i] > 0) break;
      if (i > 0 && nums[i] == nums[i - 1]) continue;
      int lo = i + 1, hi = nums.length - 1;
      int sum;
      while (lo < hi) {
        sum = nums[lo] + nums[hi];
        if (sum < -nums[i]) lo++;
        else if (sum > -nums[i]) hi--;
        else {
          result.add(Arrays.asList(nums[i], nums[lo], nums[hi]));
          while (++lo < hi && nums[lo] == nums[lo - 1]);
          while (lo < --hi && nums[hi] == nums[hi + 1]);
        }
      }
    }
    return result;
  }
}
